src/compiler/builders.js

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {buildDom as ampLayoutClassic} from '#builtins/amp-layout/amp-layout';
+
+import {buildDom as ampFitTextClassic} from '../../extensions/amp-fit-text/0.1/amp-fit-text';
+
+const builderMap = {
+  'v0': {
+    'amp-layout': ampLayoutClassic,
+  },
+  '0.1': {
+    'amp-fit-text': ampFitTextClassic,
+  },
+};
+
+/**
+ * Returns the set of component builders needed to server-render an AMP Document.
+ *
+ * @param {string} runtime
+ * @param {Array<{name: string, version: string}>} extensionList
+ * @return {Object<string, Builder>} builders
+ */
+export function getBuilders(runtime, extensionList) {
+  const builders = {...builderMap[runtime]};
+
+  for (const {name, version} of extensionList) {
+    const builder = builderMap?.[version]?.[name];
+    if (builder) {
+      builders[name] = builder;
+    }
+  }
+
+  return builders;
+}
```

Comment:
I think this code is made more confusing by treating the runtime as a special version. Instead, the `extensionList` can include an entry like `amp-layout: v0`

```suggestion
 * @param {Array<{name: string, version: string}>} extensionList
 * @return {Object<string, Builder>} builders
 */
export function getBuilders(runtime, extensionList) {
  return Object.fromEntries(extensionList
    .map(([name, version]) => [name, builderMap?.[version]?.[name]])
    .filter([, builder] => builder)
  );
}
```

Reply:
The confusing piece is going to be somewhere. Its just whether it exists here or in C++ right?

Reply:
Does it need to be handled any differently? The only thing runtime is being used for is to key into `builderMap`, which suggests it's the same as any other version string

Reply:
Builtins are different in that they are bundled with `v0.js`. That means the builder map needs to be a merging of the extension scripts in the head and the 3 `builtins` based on runtime version.

That merging will have to happen in this step or in the phase that calls into the `bento-compiler`

Reply:
But at the point this function runs, `builderMap` already includes the builtins, so they don't need to be treated any differently within this function, which can be called with ex `{'amp-layout': 'v0', 'amp-img': 'v0', ...extensionList}`

Reply:
Followed up offline. Updated.